src/module_library/secant_method.h

```diff
@@ -0,0 +1,180 @@
+ * but the intersection is not transveral, \f$f'(x) = 0\f$ at the root.
+ * This pathology slows the method.
+ *
+ * The secant method uses the same logic as Newton's method but instead of
+ * using the first derivative, we approximate \f$f\f$ using a secant, computed
+ * from two points.
+ *
+ *  \f[ f(x) \approx f(x_1) + \left(\frac{f(x_1) - f(x_0)}{x_1 - x_0}\right) \left(x -x_1\right) \f]
+ *
+ * The root of this approximation is computed to produce to refine a guess for
+ * the root. Let \f$y_n = f(x_n)\f$; the secant method's update rule is:
+ *
+ * \f[ x_{n+1} = x_n - y_n \left(\frac{x_n - x_{n-1}}{y_n - y_{n-1}}\right)\f]
+ *
+ * This formula can be rearranged into
+ * Under ideal conditions, this method approaches the rate of convergence
+ * achieved in Newton's method. The same pathologies can also frustrate it.
+ *
+ * The secant method requires two starting guesses rather than one.
+ * Good starting guesses are important to achieving good performance. It is also
+ * important to identify potential pathologies.
+ *
+ * The order of initial guesses can affect convergence, so make the second guess
+ * the closer of the two if possible.
+ *
+ * @param [in] fun A function or object with `double operator()(double)` signature
+ *             This routine is templated so any type with a call operator with
+ *              the correct type signature will work.
+ * @param [in] x0 The first guess.
+ * @param [in] x1 The second guess.
+ *             type signature secant method calls this.
+ * @param [in] par The parameters for tolerances, maximum number of iterations, etc.
+ *              See the `secant_parameters` documentation. These are passed by
+ *             reference so the secant method routine can save the final function
+ *              evaluation (as a check on accuracy) and the number of iterations.
+ *
+ * @return Approximate root. If the method fails, it will return it's last (and
+ *      presumably best) guess for the root. If `check` member field in the
+ *      secant_parameters object is not zero, then the method has failed.
+ *      This routine will terminate if NaN or non-finite guesses are encountered.
+ */
+template <typename F>
+double find_root_secant_method(
+    F fun, double x0, double x1, secant_parameters& par)
+{
+    // check inputs for trivial solutions
+    double y0 = fun(x0);
+    if (is_zero(y0, par.atol)) {
+        par.counter = 0;
+        par.check = y0;
+        return x0;
+    }
+
+    // guesses should be different
+    if (is_close(x1, x0, par.atol, par.rtol)) {
+        throw std::runtime_error(
+            "Within tolerance, x0 == x1. Initial guesses should be distinct. ");
+    }
+
+    // check second solution
+    double y1 = fun(x1);
+    if (is_zero(y1, par.atol)) {
+        par.counter = 0;
+        par.check = y1;
+        return x1;
+    }
+
+    for (size_t n = 1; n <= par.max_iter; ++n) {
+        double secant_slope = (y1 - y0) / (x1 - x0);
+
+        // SECANT UPDATE FORMULA;
```

Comment:
I also think that all caps is not appropriate, except for the usual cases in English. All caps names for constants are OK too.